### PR TITLE
Fix consensus not proceeding and acceptCheckpoint thread safety

### PR DESCRIPTION
**Pull request by wkoszycki**
this fixes unit tests and e2e test misalignment.

- consensus not proceeding when 1 node doesn't have transactions
- consensus not finished even round timeout occurred 
- concurrent block acceptance causes invalid metrics (possible misalignment as well ?)

### CODE REVIEW

File: src/main/scala/org/constellation/consensus/Round.scala [hunk 1]
```diff
@@ -329,14 +335,19 @@ class Round(roundData: RoundData,
...
+      .accept(cache)
+      .map { _ =>
+        broadcastSignedBlockToNonFacilitators(FinishedCheckpoint(cache, finalFacilitators))
+        Option(checkpointBlock)
+      }
+      .recoverWith {
+        case err =>
           log.error(s"Failed to accept majority checkpoint block due to: ${err.getMessage}", err)
-          None
+          IO.delay(None)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/consensus/Round.scala]: `IO.delay(None)` = `IO { None }`

File: src/main/scala/org/constellation/consensus/Round.scala [hunk 2]
```diff
@@ -329,14 +335,19 @@ class Round(roundData: RoundData,
       log.debug(s"[${dao.id.short}] accepting majority checkpoint block ${checkpointBlock.baseHash}  " +
         s" with txs ${checkpointBlock.transactions.map(_.hash)} " +
         s"proposed by ${sameBlocks.head._1.id.short} other blocks ${sameBlocks.size} in round ${roundData.roundId}")
-      dao.threadSafeSnapshotService.accept(cache) match {
-        case Success(_) =>
-          broadcastSignedBlockToNonFacilitators(FinishedCheckpoint(cache, finalFacilitators))
-          Some(checkpointBlock)
-        case Failure(err) =>
+      dao.threadSafeSnapshotService
+      .accept(cache)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/consensus/Round.scala]: indent

File: src/main/scala/org/constellation/consensus/Round.scala [hunk 3]
```diff
@@ -329,14 +335,19 @@ class Round(roundData: RoundData,
...
+      .map { _ =>
+        broadcastSignedBlockToNonFacilitators(FinishedCheckpoint(cache, finalFacilitators))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/consensus/Round.scala]:
```
.flatTap(_ => IO { broadcastSignedBlockToNonFacilitators(...)) })
.map(_ => Option(checkpointBlock)) // Option? or just Some(checkpointBlock)?
```

File: src/main/scala/org/constellation/consensus/Round.scala [hunk 4]
```diff
@@ -419,7 +430,7 @@ object Round {
     messages: Seq[ChannelMessage]
   )
 
-  case class StopBlockCreationRound(roundId: RoundId, maybeCB: Option[CheckpointBlock])
+  case class StopBlockCreationRound(roundId: RoundId, maybeCB: Option[CheckpointBlock],timeout: Boolean = false)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/consensus/Round.scala]: space missing

File: src/main/scala/org/constellation/primitives/EdgeDAO.scala [hunk 5]
```diff
@@ -351,42 +354,164 @@ class ThreadSafeSnapshotService(concurrentTipService: ConcurrentTipService) {
...
+          metricKey = if (maybeHeight.isEmpty) Metrics.heightEmpty else Metrics.heightNonEmpty
+          _ <- dao.metrics.incrementMetricAsync(metricKey)
+          _ <- dao.checkpointService.memPool.put(cb.baseHash, checkpoint.copy(height = maybeHeight))
+          _ <- IO.delay(dao.recentBlockTracker.put(checkpoint.copy(height = maybeHeight)))
+          _ <- acceptMessages(cb)
+          _ <- acceptTransactions(cb)
+          _ = logger.info(s"[${dao.id.short}] Accept checkpoint=${cb.baseHash}]")
+          _ <- concurrentTipService.update(cb)
+          _ <- updateAcceptedCBSinceSnapshot(cb)
+          _ <- IO.shift *> dao.metrics.incrementMetricAsync(Metrics.checkpointAccepted)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/primitives/EdgeDAO.scala]: why shifted?

File: src/test/scala/org/constellation/consensus/RoundTest.scala [hunk 6]
```diff
@@ -95,7 +95,7 @@ class RoundTest
   dao.metrics shouldReturn metrics
 
   dao.threadSafeSnapshotService shouldReturn mock[ThreadSafeSnapshotService]
-  dao.threadSafeSnapshotService.accept(*) shouldAnswer ((a: CheckpointCache) => Success(()))
+  dao.threadSafeSnapshotService.accept(*) shouldAnswer ((a: CheckpointCache) => IO.pure(()))
```
Comment (marcinwadon) [inline src/test/scala/org/constellation/consensus/RoundTest.scala]: `IO.pure(())` -> `IO.unit`

File: src/main/scala/org/constellation/primitives/EdgeDAO.scala [hunk 7]
```diff
@@ -291,6 +291,8 @@ class ThreadSafeSnapshotService(concurrentTipService: ConcurrentTipService) {
                 cbs.map(_.flatMap(_.transactions.toList.flatMap(t => List(t.src, t.dst))).toSet)
 
               logger.info("--------- Snapshot - just before lock for snapshot")
+              val baseHashesInSnap = cbData.flatMap(x => x.map(c => c.checkpointBlock.get.baseHash))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/primitives/EdgeDAO.scala]: `cbData.flatMap(_.map(_.checkpointBlock.get.baseHash))`

File: src/main/scala/org/constellation/consensus/EdgeProcessor.scala [hunk 8]
```diff
@@ -442,7 +344,10 @@ object Snapshot {
     val sortedHashes = snapshotHashes().sortBy(Distance.calculate(_, dao.id))
     sortedHashes
       .slice(ConfigUtil.snapshotClosestFractionSize, sortedHashes.size)
-      .foreach(snapId => Files.delete(Paths.get(dao.snapshotPath.pathAsString, snapId)))
+      .foreach{snapId =>
+        println(s"removing old snapshots:${snapId}")
+        Files.delete(Paths.get(dao.snapshotPath.pathAsString, snapId))
+  }
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/consensus/EdgeProcessor.scala]: and formatting

File: src/main/scala/org/constellation/consensus/Round.scala (hunk 3 above)
Comment (wkoszycki) [reply src/main/scala/org/constellation/consensus/Round.scala]: compiler is grumbling then

**Pull request merged**